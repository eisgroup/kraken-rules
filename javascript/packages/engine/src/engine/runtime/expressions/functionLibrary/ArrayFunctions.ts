/*
 *  Copyright 2018 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

export const arrayFunctions = {
    Flat,
    Count,
    Join
};

function Flat(array: any): any[] | undefined {
    if (isUndefined(array)) {
        return undefined;
    }
    if (Array.isArray(array)) {
        let flattened: any[] = [];
        for (const element of array) {
            if (Array.isArray(element)) {
                flattened = flattened.concat(element);
            } else {
                flattened.push(element);
            }
        }
        return flattened;
    }
    throw new Error("Flat function accepts only ArrayLike objects");
}

function Count(array: any): number {
    if (Array.isArray(array)) {
        return array.length;
    }
    if (isUndefined(array)) {
        return 0;
    }
    return 1;
}

/**
 *
 * @param a
 * @param b
 * @return joins two arrays into single array by preserving order and duplicates
 */
function Join(a: any[] | null | undefined, b: any[] | null | undefined): any[] {
    return [...(a || []), ...(b || [])];
}

function isUndefined(array: any): boolean {
    // tslint:disable-next-line
    return array == undefined;
}
