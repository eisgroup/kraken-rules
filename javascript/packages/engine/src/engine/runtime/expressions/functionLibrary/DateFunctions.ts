/*
 *  Copyright 2018 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import moment from "moment";
import { message } from "./function.utils";
// tslint:disable: triple-equals

function Today(): Date {
    const now = new Date();
    return new Date(now.getFullYear(), now.getMonth(), now.getDate(), 0, 0);
}

function Now(): Date {
    return new Date();
}

function FxDate(dateString?: string): Date;
function FxDate(year?: number, month?: number, day?: number): Date;

function FxDate(dateString?: string | number, month?: number, day?: number): Date {
    if (!dateString) {
        throw new Error("Failed to execute function 'Date' with parameters: " + [...arguments].join());
    }
    if (typeof dateString === "string") {
        return new Date(dateString);
    }
    if (month && day) {
        return new Date(dateString, month - 1, day);
    }
    throw new Error("Failed to execute function 'Date' with parameters: " + [...arguments].join());

}

function PlusYears(dateArg?: Date, num?: number): Date {
    if (!dateArg) {
        throw new Error(message("PlusYears", message.reason.firstParam));
    }
    if (num == undefined) {
        throw new Error(message("PlusYears", message.reason.secondParam));
    }
    return new Date(
        dateArg.getFullYear() + num, dateArg.getMonth(),
        dateArg.getDate()
    );
}

function PlusMonths(dateArg?: Date, num?: number): Date {
    if (!dateArg) {
        throw new Error(message("PlusMonths", message.reason.firstParam));
    }
    if (num == undefined) {
        throw new Error(message("PlusMonths", message.reason.secondParam));
    }
    return new Date(
        dateArg.getFullYear(),
        dateArg.getMonth() + num,
        dateArg.getDate(),
        dateArg.getHours(),
        dateArg.getMinutes(),
        dateArg.getMilliseconds()
    );
}

function PlusDays(dateArg?: Date, num?: number): Date {
    if (!dateArg) {
        throw new Error(message("PlusDays", message.reason.firstParam));
    }
    if (num == undefined) {
        throw new Error(message("PlusDays", message.reason.secondParam));
    }
    return new Date(
        dateArg.getFullYear(),
        dateArg.getMonth(),
        dateArg.getDate() + num,
        dateArg.getHours(),
        dateArg.getMinutes(),
        dateArg.getMilliseconds()
    );
}

function AsDate(dateArg?: Date): Date {
    if (!dateArg) {
        throw new Error(message("AsDate", message.reason.firstParam));
    }
    return new Date(
        dateArg.getFullYear(), dateArg.getMonth(), dateArg.getDate()
    );
}

function AsTime(dateArg?: Date): Date {
    if (!dateArg) {
        throw new Error(message("AsTime", message.reason.firstParam));
    }
    return new Date(
        dateArg.getFullYear(), dateArg.getMonth(), dateArg.getDate(), 0, 0, 0, 0
    );
}

function IsDateBetween(dateToCheck?: Date, start?: Date, end?: Date): boolean {
    if (!dateToCheck) {
        throw new Error(message("NumberOfDaysBetween", message.reason.firstParam));
    }
    if (!start) {
        throw new Error(message("NumberOfDaysBetween", message.reason.secondParam));
    }
    if (!end) {
        throw new Error(message("NumberOfDaysBetween", message.reason.thirdParam));
    }
    return dateAfter(dateToCheck, start) && dateBefore(dateToCheck, end);
}

function GetDay(date?: Date): number {
    if (!date) {
        throw new Error("Failed to execute function 'GetDay'. Parameter is absent");
    }
    try {
        return date.getDate();
    } catch (error) {
        throw new Error(`Failed to execute function 'GetDay'. Parameter '${date}' is invalid`);
    }
}

function GetYear(date?: Date): number {
    if (!date) {
        throw new Error("Failed to execute function 'GetYear'. Parameter is absent");
    }
    try {
        return date.getFullYear();
    } catch (error) {
        throw new Error(`Failed to execute function 'GetYear'. Parameter '${date}' is invalid`);
    }
}

function GetMonth(date?: Date): number {
    if (!date) {
        throw new Error("Failed to execute function 'GetMonth'. Parameter is absent");
    }
    try {
        return date.getMonth() + 1;
    } catch (error) {
        throw new Error(`Failed to execute function 'GetMonth'. Parameter '${date}' is invalid`);
    }
}

function NumberOfDaysBetween(start?: Date, end?: Date): number {
    if (!start) {
        throw new Error(message("NumberOfDaysBetween", message.reason.firstParam));
    }
    if (!end) {
        throw new Error(message("NumberOfDaysBetween", message.reason.secondParam));
    }
    const day = 24 * 60 * 60 * 1000;
    const diffTime = Math.abs((start.getTime() - end.getTime()));
    return Math.floor(diffTime / day);
}

function NumberOfMonthsBetween(d1?: Date, d2?: Date): number {
    if (!d1) {
        throw new Error(message("NumberOfMonthsBetween", message.reason.firstParam));
    }
    if (!d2) {
        throw new Error(message("NumberOfMonthsBetween", message.reason.secondParam));
    }
    try {
        const date1 = AsDate(d1);
        const date2 = AsDate(d2);

        const from = date1 < date2 ? date1 : date2;
        const to = date2 > date1 ? date2 : date1;

        let fullMonths = (to.getFullYear() - from.getFullYear()) * 12 + (to.getMonth() - from.getMonth());
        if (fullMonths > 0) {
            fullMonths -= 1;
        }

        let partialMonth = 0;
        if ((from.getFullYear() < to.getFullYear() || from.getMonth() < to.getMonth())
            && to.getDate() >= from.getDate()) {
            partialMonth = 1;
        }

        return fullMonths + partialMonth;
    } catch (error) {
        throw new Error(error("NumberOfMonthsBetween", "Parameters are invalid: " + [...arguments]));
    }
}

function NumberOfYearsBetween(start?: Date, end?: Date): number {
    return Math.floor(NumberOfMonthsBetween(start, end) / 12);
}

function Format(date?: Date, format: string = "YYYY-MM-DD"): string {
    if (!date) {
        throw new Error(message("Format", message.reason.firstParam));
    }
    return moment(date).format(format);
}

export const dateFunctions = {
    Format,
    GetYear,
    GetMonth,
    GetDay,
    Today,
    Now,
    Date: FxDate,
    DateTime: FxDate,
    PlusYears,
    PlusMonths,
    PlusDays,
    AsDate,
    AsTime,
    NumberOfDaysBetween,
    NumberOfMonthsBetween,
    NumberOfYearsBetween,
    IsDateBetween
};

const dateAfter = (date: Date, start: Date) => date >= start;
const dateBefore = (date: Date, end: Date) => date <= end;
