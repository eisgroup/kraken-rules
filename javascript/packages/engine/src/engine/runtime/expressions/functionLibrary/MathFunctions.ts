/*
 *  Copyright 2018 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
import { Numbers } from './../math/Numbers'
import { message } from '../functionLibrary/function.utils'
import { Moneys } from '../math/Moneys'

export const mathFunctions = {
    Sum,
    Avg,
    Min,
    Max,
    Round,
    RoundEven,
    Floor,
    Ceil,
    Abs,
    Sign,
    Sqrt,
    NumberSequence,
}

/**
 * Rounds to scale using IEEE 754 Round Half Up strategy.
 * <p/>
 * A Half Up strategy:
 * if the number falls midway,
 * then it is rounded to the nearest value above (for positive numbers) or below (for negative numbers)
 *
 * @param n
 * @param scale optional parameter, if not provided then will round to Integer
 * @return rounded number
 */
function Round(n: unknown, scale?: number): number {
    return Numbers.round(toNumber(n), scale)
}

/**
 * Rounds to scale using IEEE 754 Round Half Even strategy.
 * <p/>
 * A Half Even strategy:
 * if the number falls midway, then it is rounded to the nearest value with an even least significant digit
 *
 * @param n
 * @param scale optional parameter, if not provided then will round to Integer
 * @return rounded number
 */
function RoundEven(n: unknown, scale?: number): number {
    return Numbers.roundEven(toNumber(n), scale)
}

/**
 * @param n
 * @return  result of a mathematical floor function which returns
 *          greatest integer that is less than or equal to provided number
 */
function Floor(n: unknown): number {
    return Numbers.floor(toNumber(n))
}

/**
 * @param n
 * @return  result of a mathematical ceiling function which returns
 *          the least integer that is greater than or equal to provided number
 */
function Ceil(n: unknown): number {
    return Numbers.ceil(toNumber(n))
}
/**
 *
 * @param n
 * @return absolute number
 */
function Abs(n: unknown): number {
    return Numbers.abs(toNumber(n))
}

/**
 *
 * @param n
 * @return  a result of a mathematical signum function, which returns:
 *          <ul>
 *              <li>-1 - if number is less than zero</li>
 *              <li> 0 - if number is equal to zero</li>
 *              <li>+1 - if number is greater than zero</li>
 *          </ul>
 */
function Sign(n: unknown): number {
    return Numbers.sign(toNumber(n))
}

/**
 *
 * @param n
 * @return square root
 */
function Sqrt(n: unknown): number {
    return Numbers.sqrt(toNumber(n))
}

/**
 *
 * @param array that contains numbers
 * @return  average over all numbers in collection.
 *          If collection contains `null` value then rule will be ignored due to missing data.
 *          If collection is null then null is returned.
 */
function Avg(array: unknown): number | undefined {
    if (isUndefined(array) || isEmpty(array)) {
        return undefined
    }
    if (Array.isArray(array)) {
        return Numbers.avgOfArray(array.map(toNumber))
    }
    throw new Error('Avg function accepts only ArrayLike objects')
}

/**
 *
 * @param array that contains numbers
 * @return  sum over all numbers in collection.
 *          If collection contains `null` value then rule will be ignored due to missing data.
 *          If collection is null then null is returned.
 */
function Sum(array: unknown): number | undefined {
    if (isUndefined(array) || isEmpty(array)) {
        return undefined
    }
    if (Array.isArray(array)) {
        return Numbers.sumOfArray(array.map(toNumber))
    }
    throw new Error('Sum function accepts only ArrayLike objects')
}

/**
 *
 * Returns smallest number or date.
 * Returns smallest number or date from array of numbers or dates when array is passed as a first parameter
 * and second parameter is not passed.
 * Returns smallest number or date out of two numbers or dates when both parameters are passed.
 *
 * @param first is an array of numbers or array of dates or a single number or date
 * @param second number or date
 * @return smallest number or date
 */
function Min(first: unknown, second?: unknown): number | Date | undefined {
    if (arguments.length === 1) {
        return MinArray(first)
    }
    if (isNumber(first) && isNumber(second)) {
        return Numbers.min(toNumber(first), toNumber(second))
    }
    if (isDate(first) && isDate(second)) {
        return first < second ? first : second
    }
    throw new Error(
        `Function 'Min' parameters are invalid: ${first}, ${second}. Parameters must be only numbers or only dates.`,
    )
}

/**
 *
 * Returns largest number or date.
 * Returns largest number or date from array of numbers or dates when array is passed as a first parameter
 * and second parameter is not passed.
 * Returns largest number or date out of two numbers or dates when both parameters are passed.
 *
 * @param first is an array of numbers or array of dates or a single number or date
 * @param second number or date
 * @return largest number or date
 */
function Max(first: unknown, second?: unknown): number | Date | undefined {
    if (arguments.length === 1) {
        return MaxArray(first)
    }
    if (isNumber(first) && isNumber(second)) {
        return Numbers.max(toNumber(first), toNumber(second))
    }
    if (isDate(first) && isDate(second)) {
        return first > second ? first : second
    }
    throw new Error(
        `Function 'Max' parameters are invalid: ${first}, ${second}. Parameters must be only numbers or only dates.`,
    )
}

function MinArray(array: unknown): number | Date | undefined {
    if (isUndefined(array) || isEmpty(array)) {
        return undefined
    }
    if (Array.isArray(array)) {
        if (array.every(isNumber)) {
            return Numbers.minInArray(array.map(toNumber))
        }
        if (array.every(isDate)) {
            return minDateInArray(array)
        }
        throw new Error('Min function array parameter must contain only numbers or only dates')
    }
    throw new Error('Min function accepts only ArrayLike objects')
}

function MaxArray(array: unknown): number | Date | undefined {
    if (isUndefined(array) || isEmpty(array)) {
        return undefined
    }
    if (Array.isArray(array)) {
        if (array.every(isNumber)) {
            return Numbers.maxInArray(array.map(toNumber))
        }
        if (array.every(isDate)) {
            return maxDateInArray(array)
        }
        throw new Error('Max function array parameter must contain only numbers or only dates')
    }
    throw new Error('Max function accepts only ArrayLike objects')
}

function minDateInArray(array: Date[]): Date {
    let smallestDate = array[0]
    for (const date of array) {
        if (date < smallestDate) {
            smallestDate = date
        }
    }
    return smallestDate
}

function maxDateInArray(array: Date[]): Date {
    let largestDate = array[0]
    for (const date of array) {
        if (date > largestDate) {
            largestDate = date
        }
    }
    return largestDate
}

function isUndefined(array: unknown): boolean {
    return array == undefined
}

function isEmpty(array: unknown): boolean {
    return Array.isArray(array) && !array.length
}

function isDate(value: unknown): value is Date {
    return value instanceof Date
}

function isNumber(value: unknown): value is number {
    return Moneys.isMoney(value) || typeof value === 'number'
}

function toNumber(value: unknown): number {
    if (Moneys.isMoney(value)) {
        return value.amount as number
    } else if (typeof value === 'number') {
        return value as number
    }
    throw new Error(
        'Mathematical functions accepts only ' +
            "'Number' and/or 'Money' type values, but got value ('" +
            value +
            ' of type ' +
            typeof value +
            "') ",
    )
}

/**
 * Generates a sequence of numbers by adding 1 to starting number until it is less than
 * or equals to ending number. Returns a collection of numbers in order.
 * Parameters cannot be {@code null}.
 *
 * @param from sequence starting number
 * @param to   sequence end number
 * @param step   number to add each step
 * @returns ordered collection
 * @since 11.2
 */
function NumberSequence(from: number, to: number, step?: number): number[] {
    return Numbers.sequence(nonNull(from), nonNull(to), resolveStep(from, to, step))
}

function resolveStep(from: number, to: number, step?: number): number {
    if (step != undefined) {
        validateDirection(from, to, step)
        return step
    }
    return from > to ? -1 : 1
}

function validateDirection(from: number, to: number, step: number): void {
    const direction = to < from ? -1 : 1
    const directionOfStep = step > 0 ? 1 : -1
    if (direction !== directionOfStep) {
        throw new Error(
            message(NumberSequence.name, 'These parameters would generate an infinite sequence of numbers.'),
        )
    }
}

function nonNull(n?: number): number {
    if (n == undefined) {
        throw new Error(message(NumberSequence.name, message.reason.parameterNonNull))
    }
    return n
}
