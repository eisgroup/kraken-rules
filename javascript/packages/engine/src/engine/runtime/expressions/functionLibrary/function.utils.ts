/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

export function message(fxName: string, reason: string): string {
    return `Failed to execute function '${fxName}'. ${reason}`
}

message.reason = {
    parameterNonNull: 'parameter cannot be null',
    firstParam: 'First parameter is mandatory',
    secondParam: 'Second parameter is mandatory',
    thirdParam: 'Third parameter is mandatory',
}
