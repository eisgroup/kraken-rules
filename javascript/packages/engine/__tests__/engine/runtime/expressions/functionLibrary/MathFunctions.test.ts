/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { mathFunctions } from '../../../../../src/engine/runtime/expressions/functionLibrary/MathFunctions'
const { NumberSequence } = mathFunctions
import { mock } from '../../../../mock'

describe('Array Functions Test', () => {
    describe('Sum function', () => {
        it('should find sum of values from collection', () => {
            const result = mathFunctions.Sum([mock.toMoney(15.21), 6, 8, 9.000001, 0.685])
            expect(result).toBe(38.895001)
        })
        it('should return undefined when collection is empty', () => {
            const result = mathFunctions.Sum([])
            expect(result).toBe(undefined)
        })
        it('should return undefined when collection is undefined', () => {
            const result = mathFunctions.Sum(undefined)
            expect(result).toBe(undefined)
        })
        it('should throw on non array', () => {
            expect(() => mathFunctions.Sum(1)).toThrow()
        })
        it('should throw if has undefined', () => {
            expect(() => mathFunctions.Sum([undefined])).toThrow()
        })
        it('should throw if has wrong type', () => {
            expect(() => mathFunctions.Sum(['1'])).toThrow()
        })
    })
    describe('Avg function', () => {
        it('should find avg of values from collection', () => {
            const result = mathFunctions.Avg([mock.toMoney(12.99), 6, 18.5, 0.12, 0.6])
            expect(result).toBe(7.642)
        })
        it('should return undefined when collection is empty', () => {
            const result = mathFunctions.Min([])
            expect(result).toBe(undefined)
        })
        it('should return undefined when avg collection is undefined', () => {
            const result = mathFunctions.Avg(undefined)
            expect(result).toBe(undefined)
        })
        it('should throw on non array', () => {
            expect(() => mathFunctions.Avg(1)).toThrow()
        })
        it('should throw if has undefined', () => {
            expect(() => mathFunctions.Avg([undefined])).toThrow()
        })
        it('should throw if has wrong type', () => {
            expect(() => mathFunctions.Avg(['1'])).toThrow()
        })
    })
    describe('Round function', () => {
        it('should round to integer', () => {
            expect(mathFunctions.Round(1)).toBe(1)
            expect(mathFunctions.Round(1.5)).toBe(2)
            expect(mathFunctions.Round(2.5)).toBe(3)
            expect(mathFunctions.Round(-1)).toBe(-1)
            expect(mathFunctions.Round(-1.5)).toBe(-2)
            expect(mathFunctions.Round(-2.5)).toBe(-3)
        })
        it('should round to even integer', () => {
            expect(mathFunctions.RoundEven(1)).toBe(1)
            expect(mathFunctions.RoundEven(1.5)).toBe(2)
            expect(mathFunctions.RoundEven(2.5)).toBe(2)
            expect(mathFunctions.RoundEven(-1)).toBe(-1)
            expect(mathFunctions.RoundEven(-1.5)).toBe(-2)
            expect(mathFunctions.RoundEven(-2.5)).toBe(-2)
        })
        it('should round to scale', () => {
            expect(mathFunctions.Round(1, 2)).toBe(1)
            expect(mathFunctions.Round(1.555, 2)).toBe(1.56)
            expect(mathFunctions.Round(100.565, 2)).toBe(100.57)
            expect(mathFunctions.Round(-1, 2)).toBe(-1)
            expect(mathFunctions.Round(-1.555, 2)).toBe(-1.56)
            expect(mathFunctions.Round(-100.565, 2)).toBe(-100.57)
        })
        it('should round even to scale', () => {
            expect(mathFunctions.RoundEven(1, 2)).toBe(1)
            expect(mathFunctions.RoundEven(1.555, 2)).toBe(1.56)
            expect(mathFunctions.RoundEven(100.565, 2)).toBe(100.56)
            expect(mathFunctions.RoundEven(-1, 2)).toBe(-1)
            expect(mathFunctions.RoundEven(-1.555, 2)).toBe(-1.56)
            expect(mathFunctions.RoundEven(-100.565, 2)).toBe(-100.56)
        })
    })
    describe('Floor function', () => {
        it('should floor', () => {
            expect(mathFunctions.Floor(1)).toBe(1)
            expect(mathFunctions.Floor(1.9)).toBe(1)
            expect(mathFunctions.Floor(-1)).toBe(-1)
            expect(mathFunctions.Floor(-1.1)).toBe(-2)
        })
    })
    describe('Ceil function', () => {
        it('should ceil', () => {
            expect(mathFunctions.Ceil(1)).toBe(1)
            expect(mathFunctions.Ceil(1.1)).toBe(2)
            expect(mathFunctions.Ceil(-1)).toBe(-1)
            expect(mathFunctions.Ceil(-1.9)).toBe(-1)
        })
    })
    describe('Abs function', () => {
        it('should abs', () => {
            expect(mathFunctions.Abs(1)).toBe(1)
            expect(mathFunctions.Abs(-1.1)).toBe(1.1)
        })
    })
    describe('Sign function', () => {
        it('should sign', () => {
            expect(mathFunctions.Sign(1)).toBe(1)
            expect(mathFunctions.Sign(0)).toBe(0)
            expect(mathFunctions.Sign(-1.1)).toBe(-1)
        })
    })
    describe('Max function', () => {
        it('should return max of two numbers', () => {
            expect(mathFunctions.Max(0, 1)).toBe(1)
            expect(mathFunctions.Max(-1.1, -10)).toBe(-1.1)
            expect(() => mathFunctions.Max(0, undefined)).toThrow()
        })
        it('should return min of two dates', () => {
            expect(mathFunctions.Max(new Date('2020-01-01'), new Date('2020-01-02')) as Date).k_toBeDateEqualTo(
                new Date('2020-01-02'),
            )

            expect(() => mathFunctions.Max(0, new Date(2020, 1, 1))).toThrow()
            expect(() => mathFunctions.Max(0, undefined)).toThrow()
        })
        it('should return min of date collection', () => {
            expect(mathFunctions.Max([new Date('2020-01-01'), new Date('2020-01-02')]) as Date).k_toBeDateEqualTo(
                new Date('2020-01-02'),
            )
        })
        it('should find max of values from collection', () => {
            const result = mathFunctions.Max([mock.toMoney(19.99), 6, 13.5, 0.12, 0.6])
            expect(result).toBe(19.99)
        })
        it('should return undefined when collection is empty', () => {
            const result = mathFunctions.Max([])
            expect(result).toBe(undefined)
        })
        it('should return when collection is undefined', () => {
            const result = mathFunctions.Max(undefined)
            expect(result).toBe(undefined)
        })
        it('should throw on non array', () => {
            expect(() => mathFunctions.Max(1)).toThrow()
        })
        it('should throw if has undefined', () => {
            expect(() => mathFunctions.Max([undefined])).toThrow()
        })
        it('should throw if has wrong type', () => {
            expect(() => mathFunctions.Max(['1'])).toThrow()
        })
    })
    describe('Min function', () => {
        it('should return min of two numbers', () => {
            expect(mathFunctions.Min(0, 1)).toBe(0)
            expect(mathFunctions.Min(-1.1, -10)).toBe(-10)
            expect(() => mathFunctions.Min(0, undefined)).toThrow()
        })
        it('should return min of two dates', () => {
            expect(mathFunctions.Min(new Date('2020-01-01'), new Date('2020-01-02')) as Date).k_toBeDateEqualTo(
                new Date('2020-01-01'),
            )

            expect(() => mathFunctions.Min(0, new Date(2020, 1, 1))).toThrow()
            expect(() => mathFunctions.Min(0, undefined)).toThrow()
        })
        it('should return min of date collection', () => {
            expect(mathFunctions.Min([new Date('2020-01-01'), new Date('2020-01-02')]) as Date).k_toBeDateEqualTo(
                new Date('2020-01-01'),
            )
        })
        it('should find min of values from collection', () => {
            expect(mathFunctions.Min([mock.toMoney(15.99), 6, 18.5, 0.12, 0.6])).toBe(0.12)
            expect(mathFunctions.Min([-5, 6, 18.5, 0.12, 0.6])).toBe(-5)
        })
        it('should return undefined when collection is empty', () => {
            const result = mathFunctions.Min([])
            expect(result).toBe(undefined)
        })
        it('should return undefined when collection is undefined', () => {
            const result = mathFunctions.Min(undefined)
            expect(result).toBe(undefined)
        })
        it('should throw on non array', () => {
            expect(() => mathFunctions.Min(1)).toThrow()
        })
        it('should throw if has undefined', () => {
            expect(() => mathFunctions.Max([undefined])).toThrow()
        })
        it('should throw if has wrong type', () => {
            expect(() => mathFunctions.Max(['1'])).toThrow()
        })
    })
    describe('Sqrt function', () => {
        it('should sqrt', () => {
            expect(mathFunctions.Sqrt(1)).toBe(1)
            expect(mathFunctions.Sqrt(9)).toBe(3)
            expect(mathFunctions.Sqrt(15.29344444444444)).toBe(3.910683373074895)
            expect(() => mathFunctions.Sqrt(undefined)).toThrow()
        })
    })
    describe('Generator Functions', () => {
        describe('default steps', () => {
            describe('default steps', () => {
                it('should generate sequence', () => {
                    expect(NumberSequence(0, 3)).toMatchObject([0, 1, 2, 3])
                    expect(NumberSequence(3, 0)).toMatchObject([3, 2, 1, 0])
                    expect(NumberSequence(0, 0)).toMatchObject([0])
                    expect(NumberSequence(30, 32, 1)).toMatchObject([30, 31, 32])
                    expect(NumberSequence(0.3, 0.39, 0.01)).toMatchObject([
                        0.3, 0.31, 0.32, 0.33, 0.34, 0.35, 0.36, 0.37, 0.38, 0.39,
                    ])
                    expect(NumberSequence(0.33, 0.339, 0.001)).toMatchObject([
                        0.33, 0.331, 0.332, 0.333, 0.334, 0.335, 0.336, 0.337, 0.338, 0.339,
                    ])
                    expect(NumberSequence(0.333, 0.3339, 0.0001)).toMatchObject([
                        0.333, 0.3331, 0.3332, 0.3333, 0.3334, 0.3335, 0.3336, 0.3337, 0.3338, 0.3339,
                    ])
                    expect(NumberSequence(10002.00001, 10002.000051, 0.000001)).toMatchSnapshot(
                        '10002.00001, 10002.000051, 0.000001',
                    )
                    expect(NumberSequence(1.00001, 1.00002, 1.000001)).toMatchSnapshot('1.00001, 1.00002, 1.000001')
                    expect(NumberSequence(1000.0001, 1000.0002, 0.00001)).toMatchSnapshot(
                        '1000.0001, 1000.0002, 0.00001 - 5 digits',
                    )
                    expect(NumberSequence(1000000.00001, 1000000.00002, 0.000001)).toMatchSnapshot(
                        '1000000.00001, 1000000.00002, 0.000001 - 6 digits',
                    )
                })
                it('should throw on null parameters', () => {
                    const undefinedParameter = undefined as unknown as number
                    expect(() => NumberSequence(undefinedParameter, 3)).toThrowError(NumberSequence.name)
                    expect(() => NumberSequence(1, undefinedParameter)).toThrowError(NumberSequence.name)
                })
            })
        })
        describe('explicit steps', () => {
            it('should generate sequence', () => {
                expect(NumberSequence(0, 3, 2)).toMatchObject([0, 2])
                expect(NumberSequence(4, 0, -2)).toMatchObject([4, 2, 0])
                expect(NumberSequence(10, 11, 12)).toMatchObject([10])
            })
            it('should throw on null parameters', () => {
                const undefinedParameter = undefined as unknown as number
                expect(() => NumberSequence(undefinedParameter, 3)).toThrowError(NumberSequence.name)
                expect(() => NumberSequence(1, undefinedParameter)).toThrowError(NumberSequence.name)
            })
            it('should throw on infinity generation', () => {
                expect(() => NumberSequence(10, 5, 1)).toThrowError(NumberSequence.name)
                expect(() => NumberSequence(5, 10, -1)).toThrowError(NumberSequence.name)
                expect(() => NumberSequence(1, 2, 0)).toThrowError(NumberSequence.name)
            })
        })
    })
})
