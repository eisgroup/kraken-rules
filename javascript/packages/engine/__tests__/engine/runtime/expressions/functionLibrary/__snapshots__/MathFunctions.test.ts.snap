// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Array Functions Test Generator Functions default steps default steps should generate sequence: 1.00001, 1.00002, 1.000001 1`] = `
Array [
  1.00001,
]
`;

exports[`Array Functions Test Generator Functions default steps default steps should generate sequence: 1000.0001, 1000.0002, 0.00001 - 5 digits 1`] = `
Array [
  1000.0001,
  1000.00011,
  1000.00012,
  1000.00013,
  1000.00014,
  1000.00015,
  1000.00016,
  1000.00017,
  1000.00018,
  1000.00019,
  1000.0002,
]
`;

exports[`Array Functions Test Generator Functions default steps default steps should generate sequence: 10002.00001, 10002.000051, 0.000001 1`] = `
Array [
  10002.00001,
  10002.000011,
  10002.000012,
  10002.000013,
  10002.000014,
  10002.000015,
  10002.000016,
  10002.000017,
  10002.000018,
  10002.000019,
  10002.00002,
  10002.000021,
  10002.000022,
  10002.000023,
  10002.000024,
  10002.000025,
  10002.000026,
  10002.000027,
  10002.000028,
  10002.000029,
  10002.00003,
  10002.000031,
  10002.000032,
  10002.000033,
  10002.000034,
  10002.000035,
  10002.000036,
  10002.000037,
  10002.000038,
  10002.000039,
  10002.00004,
  10002.000041,
  10002.000042,
  10002.000043,
  10002.000044,
  10002.000045,
  10002.000046,
  10002.000047,
  10002.000048,
  10002.000049,
  10002.00005,
  10002.000051,
]
`;

exports[`Array Functions Test Generator Functions default steps default steps should generate sequence: 1000000.00001, 1000000.00002, 0.000001 - 6 digits 1`] = `
Array [
  1000000.00001,
  1000000.000011,
  1000000.000012,
  1000000.000013,
  1000000.000014,
  1000000.000015,
  1000000.000016,
  1000000.000017,
  1000000.000018,
  1000000.000019,
  1000000.00002,
]
`;
